File: C++/102-Binary_Tree_Level_Order_Traversal.cpp
// Runtime: 4 ms, faster than 93.37% of C++ online submissions for Binary Tree Level Order Traversal.
// Memory Usage: 16.1 MB, less than 5.63% of C++ online submissions for Binary Tree Level Order Traversal.
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (root == NULL)
            return res;
        vector<TreeNode*> layers = {root};
        bfs(res, layers);
        return res;
    }
private:
    vector<vector<int>> res;
    void bfs(vector<vector<int>>& res, vector<TreeNode*> layers) {
        if (layers.size() == 0)
            return;
        vector<TreeNode*> new_layers;
        vector<int> layer_value;
        for (TreeNode* node : layers) {
            if (node->left != NULL)
                new_layers.push_back(node->left);
            if (node->right != NULL)
                new_layers.push_back(node->right);
            layer_value.push_back(node->val);
        }
        res.push_back(layer_value);
        bfs(res, new_layers);
    }
};

File: Algorithms/Python3.x/102-Binary_Tree_Level_Order_Traversal.py
# bfs 队列
# Runtime: 40 ms, faster than 98.37% of Python3 online submissions for Binary Tree Level Order Traversal.
# Memory Usage: 12.5 MB, less than 0.86% of Python3 online submissions for Binary Tree Level Order Traversal.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: 'TreeNode') -> 'List[List[int]]':
        if not root: return []
        queue, result = [], []
        queue.append(root)
        while len(queue) > 0:
            line, tmp = [], []
            for node in queue:
                line.append(node.val)
                if node.left: tmp.append(node.left)
                if node.right: tmp.append(node.right)
            queue = tmp
            result.append(line)
        return result
